### PR TITLE
Integrating RH Service Mesh with Kubeflow

### PR DESCRIPTION
**Pull request by nakfour**
This Pull request will include all the necessary changes needed to integrate RH Mesh Service (installed from Operator Catalog) with Kubeflow 0.7
Solving Issue: https://gitlab.com/opendatahub/opendatahub-operator/issues/116
To test:
1. Install RH Service Mesh by following instructions: https://access.redhat.com/documentation/en-us/openshift_container_platform/4.2/html/service_mesh/service-mesh-installation#installing-ossm. Make sure to wait for each operator install to be successful before moving to the next step. 
2. git clone the kubeflow manifest repo
3. build and apply with config file kfctl_openshift-servicemesh.yaml
3. For testing if the integration succeeded
     a. oc get virtualservices
     b. Access the portals using http://<istio ingress url>/<path of component from virtualservice>
4. Also we can test by launching a notebook server in a specific namespace. However this namespace has to be added to the Service Mesh Member Roll instance called "default" in the istio-system namespace.

### CODE REVIEW

**Comment by crobby** [comment]
So far, this is installing ok.  The only oddity I've seen so far is with respect to http vs https.  http seems to work, whereas https isn't working.
Might be good to build this on top of the changes in #4 to have as many working components as possible.

**Comment by vpavlin** [comment]
I will retest on a fresh cluster tomorrow - I did something wrong and had to delete the `istio-system` namespace and recreate the deployment there to get it working, but the second attempt went well.

@nakfour Can you please explain the changes you are making in this PR? I have hard time understanding why they are needed, because when I diff the resources I see

* no changes in the cluster-role
* parameter rename in the kf-istio-resources

```
$ diff -u istio/istio/base/kf-istio-resources.yaml istio/service-mesh/base/kf-servicemesh-resources.yaml 
--- istio/istio/base/kf-istio-resources.yaml	2020-01-14 17:24:16.974560873 +0100
+++ istio/service-mesh/base/kf-servicemesh-resources.yaml	2020-01-14 17:24:16.975560883 +0100
@@ -108,4 +108,4 @@
   name: default
   namespace: istio-system
 spec:
-  mode: $(clusterRbacConfig)
+  mode: $(serviceMeshRbacConfig)
```

So it seems unnecessary to add a specific servicemesh subdir, simply comment out the istio-crds and istio-install in the kfdef...or what em I missing?

**Comment by vpavlin** [comment]
Oh, I see - you actually changed the one thing that needed a change in the istio base as well, that's what confused me.

We can do this with openshift overlay without needing to copy all the other files around!:)

If you remove the changes you made and add the following to `istio/istio/overlays/servicemesh`

`default-ClusterRbacConfig.yaml`
```

bases:
- ../../base

patchesJson6902:
- target:
    kind: ClusterRbacConfig
    name: default
    version: v1alpha1
    group: rbac.istio.io
  path: default-ClusterRbacConfig.yaml

```

`kustomization.yaml`

```
bases:
- ../../base

patchesJson6902:
- target:
    kind: ClusterRbacConfig
    name: default
    version: v1alpha1
    group: rbac.istio.io
  path: default-ClusterRbacConfig.yaml

```

And also add 

```
- path: spec/mode
  kind: ServiceMeshRbacConfig

```

to `istio/istio/base/params.yaml` (this is a bit meh, but harmless)

And just use the istio/istio component with 

```
overlays:
- servicemesh
```

Guess I can prepare a patch or something if this is too confusing:D

**Comment by nakfour** [comment]
@vpavlin the change I made is not really an overlay. In the changes you suggested, where is the ServiceMeshRbac resource defined?

**Comment by nakfour** [comment]
@crobby the kubeflow gateway installed in only for HTTP. To get HTTPS we need certificates to add to istio, a kybeflow gateway with HTTPS support and the virtual service has to specify that too

**Comment by crobby** [comment]
> @crobby the kubeflow gateway installed in only for HTTP. To get HTTPS we need certificates to add to istio, a kybeflow gateway with HTTPS support and the virtual service has to specify that too

@nakfour Makes total sense.  I just re-verified against the non-service mesh install and discovered the exact same thing.  Thanks for checking.

**Comment by crobby** [comment]
I will need to re-verify this, but I just tried importing a pipeline via URL and it failed.  It's giving me a 504--gateway timeout, but the URL I'm attempting to import is responding fine.  Can someone else try pipeline import?

**Comment by vpavlin** [comment]
I was able to start a notebook server when I added the `vpavlin` namespace to the member roll, but I cannot access it and get error `Sorry, /notebook/vpavlin/test/ is not a valid page`

**Comment by crobby** [comment]
Other things that don't seem to work for me with this install:  
admission-webhook-bootstrap-stateful-set-0  --- pod dies  (do we need this? I think maybe not)
minio -- pod dies (Likely that we don't need/want minio)
ml-pipelinepersistenceagent  -- pod dies

**Comment by nakfour** [comment]
@vpavlin I did not get the same error as you, with page not found. The url I have for the notebook is http://<istio ingress>/notebook/<profilenamespace>/<notebookname>/tree?
Did you click on the "connect" button on the main portal to get to the notebook?

**Comment by nakfour** [comment]
@crobby the provided kfdef config file with this PR serves as an example only, I think we need to address each Kubeflow components (why it is crashing, other than istio issues) separately. But I would definitly like to come up with a tight kfdef config file for Openshift.

**Comment by vpavlin** [comment]
> @vpavlin I did not get the same error as you, with page not found. The url I have for the notebook is http://"istio ingress"/notebook/"profilenamespace"/"notebookname"/tree?
> Did you click on the "connect" button on the main portal to get to the notebook?

Yes, I clicked the `connect` button and also tried to construct the URL myself based on your description - same result..I am going to try to clean up my cluster

**Comment by vpavlin** [comment]
This error 

```
create Pod test-0 in StatefulSet test failed error: Internal error occurred: admission plugin "MutatingAdmissionWebhook" failed to complete mutation in 13s
```

seems to be caused by `katib-mutating-webhook-config` - when I deleted the webhook, the pod came up

**Comment by vpavlin** [comment]
Ok, so I got notebooks working following the how-to. 2 things I had to do:

1. delete the katib webhook - not good, I am going to dig deeper to figure out what's wrong
2. the notebook virtual service started working when I added my profile namespace to ServiceMesh MemberRoll - it would not pick up the virtualservice without it - but I was able to add it after it was created - this might be actually a problem as we would need to have to enhance the profile controller

**Comment by vpavlin** [comment]
@sabre1041 PTAL if you have any ideas:-)

**Comment by nakfour** [comment]
@crobby @vpavlin @LaVLaS  added a network policy that solved the notebook pod issue with katib mutating webhook. Please review. 
PS: At this point the network policy allows all ingress traffic to namespace kubeflow, which basically was the case before installing RH Service Mesh.

File: istio/service-mesh/base/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Manifest for Service Mesh in Kubeflow
+
+- `install` dir contains the manifest to install Service Mesh
```
Comment (vpavlin) [inline istio/service-mesh/base/README.md]: Where is the `install` dir?

**Comment by vpavlin** [comment]
What is the point of having Service Mesh installed in that case?

Can we try to investigate further if there is a better way to handle this than add network policy?

**Comment by crobby** [comment]
Ok, I've tested this out and it seems to work.  
It does seem really awkward to have to add each profile/project/namespace into the member roll.  Is any real user likely to want to do that?

**Comment by vpavlin** [comment]
I think we will need to figure out an automated way how to update that - no idea what is the best way to do it, yet:)

Also, I think we should not use this "allow all" networkpolicy, but at least use what provided the Service Mesh team and label katib so that only Katib has the port open and not every single pod

**Comment by nakfour** [comment]
@crobby @vpavlin we probably can edit the member role and add the namespace to it, however this will require cluster admin. Does the KF controller that creates the user profile/namespace have cluster admin rights?

**Comment by crobby** [comment]
@nakfour The profile controller SA does seem to get a clusterrolebinding bound to cluster-admin

**Comment by nakfour** [comment]
@crobby @vpavlin I restricted  the  network policy to just the katib controller on 443 port. And I added the default member roll creation as part of the service mesh install. It had to be in its own folder/kustomize item since it needs to deploy the CR in the istio-system namespace. One more thing is left to add code to the profile controller to auto add the name of the namespace to this created memberroll  CR. I am suggesting we merge this PR than add that part later since it will require a significant amount of work, the api to change a CR is not simple using the GO client library.

File: kfdef/kfctl_openshift-servicemesh.yaml [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        path: profiles
+    name: profiles
+  - kustomizeConfig:
+      repoRef:
+        name: manifests
+        path: seldon/seldon-core-operator
+    name: seldon-core-operator
+  repos:
+  - name: manifests
+    uri: /Users/jnakfour/kubeflow/odh-kubeflow-fork/service-mesh-odh-fork/manifests
```
Comment (vpavlin) [inline kfdef/kfctl_openshift-servicemesh.yaml]: Please change this

**Comment by vpavlin** [comment]
I tested this and it worked fine - I only verified that everything is running and I can create and connect to a Jupyter notebook pod

Few thoughts

* Are you going to consider my suggestions here https://github.com/opendatahub-io/manifests/pull/8#issuecomment-574307550 ? (We can also merge this and I can submit a follow up PR
* Why does the member roll need to be a separate subfolder? Can it be just an overlay?
* There is still this question about `install` dir (https://github.com/opendatahub-io/manifests/pull/8/files#r374618446)

And one general thought - is this repository the right place to have this change? Should we rather make a decision on the naming for ODH manifests repositories and move it there - this sounds more like something ODH would provide - switch from Istio upstream to RH supported Service Mesh, than what would go into Kubeflow upstream

**Comment by crobby** [comment]
> I tested this and it worked fine - I only verified that everything is running and I can create and connect to a Jupyter notebook pod
> 
> Few thoughts
> 
>     * Are you going to consider my suggestions here [#8 (comment)](https://github.com/opendatahub-io/manifests/pull/8#issuecomment-574307550) ? (We can also merge this and I can submit a follow up PR
> 
>     * Why does the member roll need to be a separate subfolder? Can it be just an overlay?
> 
>     * There is still this question about `install` dir (https://github.com/opendatahub-io/manifests/pull/8/files#r374618446)
> 
> 
> And one general thought - is this repository the right place to have this change? Should we rather make a decision on the naming for ODH manifests repositories and move it there - this sounds more like something ODH would provide - switch from Istio upstream to RH supported Service Mesh, than what would go into Kubeflow upstream

Good point about this being maybe more of an ODH thing than an upstream KF thing.  Ideally, we just have one upstream set of manifests/kfdef to maintain.  This might be the perfect example of a downstream addition.

**Comment by nakfour** [comment]
I actually disagree with @vpavlin and @crobby with regards to this not being an upstream Kubeflow change. I think we do have users of RH service mesh that want to install Kubeflow and now they have the option to do so if we submit this upstream. Kubefow in the documentation specifically mention the ability to install Kiubeflow with already installed istio. Its the exact same changes we have to support Openshift, we have users that need the OCP changes and we also have users that need the RH Service mesh + Kubeflow.

File: istio/service-mesh/base/README.md (hunk 1 above)
Comment (nakfour) [reply istio/service-mesh/base/README.md]: this was inherited, deleted it

**Comment by nakfour** [comment]
@vpavlin answers to your questions.
1. Are you going to consider my suggestions here [#8 (comment)](https://github.com/opendatahub-io/manifests/pull/8#issuecomment-574307550) ? (We can also merge this and I can submit a follow up PR
I would like to know what are the technical benefits of using an overlay instead of what I have here. I just followed what the Kubeflow community used, they did not use an overlay. 
2. Why does the member roll need to be a separate subfolder? Can it be just an overlay? Please read my previous comment (https://github.com/opendatahub-io/manifests/pull/8#issuecomment-585001712) I specifically pointed out that it needs to install in a different namespace than kubeflow. I could not get kustomize to install different resources in different namespaces, it was always overwritten by the one namespace, either kubeflow or istio-system. Do you know how to do it? Basically one kustomize that installs resources in two namespaces

**Comment by vpavlin** [comment]
> what are the technical benefits of using an overlay instead of what I have here.

The technical benefit, as I mentioned in the original comment is that we can do this "without needing to copy all the other files around" - the goal is to minimize maintenance burden - if KF changes something in istio deployments you might need to bring it to your copy of the files to keep other components working - so patching is in general better than copying - it is similar to when you write a program code, you don't want to copy functions definition because of one line being different but rather abstract that part out.

> I specifically pointed out that it needs to install in a different namespace than kubeflow.

IIUC the different namespace than kubeflow is pretty specifically `istio-system` right? And you already have resources being deployed there, so why not to just add it to that component? (Or an overlay in case you decide to go that route)

> Kubefow in the documentation specifically mention the ability to install Kiubeflow with already installed istio

Can you please point me to that documentation? I could not find it

> RH Service mesh + Kubeflow

One thing to consider here is the distinction between istio and RH Service Mesh - istio is upstream project and RH Service Mesh is a product, this is why I think it does not necessarily belong to upstream Kubeflow and we might want to add it in ODH.

Also since there is many manual steps involved in installing Service Mesh right now, I am not a big fan of proposing it upstream in general. But if the general consensus is to do it, I will not block it

@LaVLaS @pdmack Any thoughts?

**Comment by nakfour** [comment]
@vpavlin here are my answers (I am writing this again because github tanked UGH !!)
1. I agree to use overlay for RH Service mesh instead of a copied folder so we dont have to re-copy all changes.
2. I will keep the member roll in a folder on its own as is right now, this makes more sense to me than overlaying other istio.
3. Installing Kubeflow with already installed istio is in the kfdef file see https://github.com/kubeflow/manifests/blob/v0.7-branch/kfdef/kfctl_k8s_istio.0.7.1.yaml specifically: # Istio install. If not needed, comment out istio-crds and istio-install.
4. I disagree that since RH service mesh is a product then we dont need to upstream. I can have the same argument for Openshift, it is a product too and we still upstream. RH Service mesh is also based on open source projects: Istio, Kiali and Jaeger same as OCP based on OKD,

**Comment by nakfour** [comment]
@vpavlin @crobby I converted the service mesh component to an overlay. Member roll is still in its own folder as suggested above because of the need to install in istio-system namespace

**Comment by crobby** [comment]
Just tried this out on my OpenShift 4.3 cluster (CRC)
When I go to do kfctl apply, I get....

```
ERRO[0000] error evaluating kustomization manifest for istio Error no matches for OriginalId rbac.istio.io_v1alpha1_ClusterRbacConfig|~X|default; no matches for CurrentId rbac.istio.io_v1alpha1_ClusterRbacConfig|~X|default; failed to find unique target for patch rbac.istio.io_v1alpha1_ClusterRbacConfig|default  filename="kustomize/kustomize.go:163"
Error: failed to apply:  (kubeflow.error): Code 500 with message: kfApp Apply failed for kustomize:  (kubeflow.error): Code 500 with message: error evaluating kustomization manifest for istio Error no matches for OriginalId rbac.istio.io_v1alpha1_ClusterRbacConfig|~X|default; no matches for CurrentId rbac.istio.io_v1alpha1_ClusterRbacConfig|~X|default; failed to find unique target for patch rbac.istio.io_v1alpha1_ClusterRbacConfig|default
```

**Comment by vpavlin** [comment]
@nakfour The error @crobby hit is because you left the changes in the file `istio/istio/base/kf-istio-resources.yaml` - please revert those and it should be all good.

Can you please rename the file `kfdef/kfctl_openshift-servicemesh.yaml` to `kfdef/kfctl_openshift_servicemesh.yaml`, so that the name is consistent with the rest?

Also, can you please rebase? (to avoid the Seldon issue on 4.3 while testing)

**Comment by vpavlin** [comment]
When you rebase, you will also need to add `cert-manager` to your KFDef file

**Comment by nakfour** [comment]
@crobby added the line you requested on crc. @LaVLaS when you have time can you please review since @vpavlin is out.

**Comment by vpavlin** [comment]
We should test other webhooks as well - they might have similar issues to Katib

**Comment by LaVLaS** [comment]
I was able to deploy RH Service Mesh and deploy the openshift service mesh kfdek file. Access to jupyter notebooks and basic pipeline access & execution worked

**Comment by tsailiming** [comment]
@LaVLaS Can you share a basic hello world pipeline?

I keep getting a  `k8sapi executor does not support outputs from base image layer. must use emptyDir` error even when I use a simple pipeline that creates a Volume

```
@dsl.pipeline(
    name="VolumeOp Basic",
    description="A Basic Example on VolumeOp Usage."
)
def volumeop_basic(size):
    vop = dsl.VolumeOp(
        name="create-pvc",
        resource_name="my-pvc",
        modes=dsl.VOLUME_MODE_RWO,
        size=size
    )

if __name__ == '__main__':
    kfp.compiler.Compiler().compile(volumeop_basic, __file__ + '.yaml')
```

**Pull request merged**